Feature Add: Lora/Embed Auto complete

ElliottLester opened the pull request:
- Lora/Embedding now suggest completions to the user.
- After typing `embedding:` or `<lora:`
- Use enter/return to use the auto complete

Looking for feedback on:
- Do we need a way to disable this?
- Any other patterns?
- Should we use an explicit shortcut to invoke (eg ctrl-e)

ElliottLester committed:
Add Lora/Embed auto complete

- Lora/Embedding now suggest completions to the user.
- Use tab to use the auto complete

ElliottLester committed:
Update Autocomplete

- Set the lists on  Construction
- use re.Match instead of WordUnderCursor

JasonS09 commented:
> Looking for feedback on:
> 
> * Shortcuts tab vs enter?
> * Do we need a way to disable this?
> * Any other patterns?
> * Should we use an explicit shortcut to invoke (eg ctrl-e)

Thanks for this PR, it's something I wanted implemented in this plugin. As for how it looks, I don't think there's nothing wrong with the current approach. Making it optional and implementing a shortcut for me seems like an overkill; if for any reason I start seeing the need, or someone requests to implement either, we'll do it then. 

Just one thing though, is there a specific reason why you decided to use tab instead of enter for shorcut? I think enter is more intuitive for the users.

ElliottLester commented:
> > Looking for feedback on:
> > 
> > * Shortcuts tab vs enter?
> > * Do we need a way to disable this?
> > * Any other patterns?
> > * Should we use an explicit shortcut to invoke (eg ctrl-e)
> 
> Thanks for this PR, it's something I wanted implemented in this plugin. As for how it looks, I don't think there's nothing wrong with the current approach. Making it optional and implementing a shortcut for me seems like an overkill; if for any reason I start seeing the need, or someone requests to implement either, we'll do it then.
> 
> Just one thing though, is there a specific reason why you decided to use tab instead of enter for shorcut? I think enter is more intuitive for the users.

TAB was the default key in the QT documentation for QCompleter and in some IDEs.
I have found myself pressing enter, so let me try and implement it and see if there's a conflict with the QPlainTextEdit

ElliottLester committed:
Update AutoComplete

- Rename auto_complete_* -> fuzzy_match_* to avoid confusion
- Change auto complete key to return & enter
- Add suffix support so lora can add weight automatically
- Add gui_test.py for faster iterations

JasonS09 commented on frontends/krita/krita_comfy/widgets/prompt.py:
```diff
@@ -1,37 +1,62 @@
+        tc.movePosition(QTextCursor.EndOfWord)
+        tc.insertText(text[-extra:])
+        self.setTextCursor(tc)
+        self.completer.hide_popup()
```
```suggestion
    def insertCompletion(self, text):
        tc = self.textCursor()
        # Delete the current word after a colon
        while tc.positionInBlock() > 0 and not tc.atBlockStart() and tc.selectedText() != ":":
            tc.deletePreviousChar()
            tc.movePosition(QTextCursor.Left, QTextCursor.KeepAnchor)   
        # Insert the new word
        tc.movePosition(QTextCursor.EndOfWord)
        tc.insertText(text)
        self.setTextCursor(tc)
        self.completer.hide_popup()
```

The current implementation successfully completes the embedding/lora name, however, it would be better if it replaced the current written name. Something like this allows users start writing names in lower case and getting the correct name of the resource.

JasonS09 commented on frontends/krita/krita_comfy/widgets/prompt.py:
```diff
@@ -1,37 +1,62 @@
+            if key == Qt.Key_Enter or key == Qt.Key_Return:
+                self.completer.insertText.emit(self.completer.getSelected())
+                return
+
```
```suggestion
            if key == Qt.Key_Left or key == Qt.Key_Right:
                return
```
We could also prevent users from moving left or right when the popup is visible. I think it's the simplest way to prevent them from messing with the whole prompt if they press enter and cursor is not in intended place. We wouldn't also have to deal with having to hide the popup if cursor moves outside of current word. However, you're welcome to try your own approaches.

ElliottLester committed:
Update AutoComplete

- Use replace not append
- support more than one match per line
- Update lora_start regex to not match spaces
- Set match mode to contains not starts with
- try catch try_auto_complete as throwing into native QT causes crash.
